### PR TITLE
Add support for arbitrary AWS accessKey/secretKey k8s secrets

### PR DESCRIPTION
**Pull request by marshields**
**Problem**

Configuration expects rigid format for AWS accessKey/secretKey.

Our internal, k8s secrets manager does not present the values needed in the required secret naming/structure.

**Solution**

Add support for arbitrary AWS accessKey/secretKey k8s secrets.

**Result**

Able to use our internal, k8s secrets manager sourced AWS access and secret keys.

### CODE REVIEW

File: pkg/api/v1beta1/awspcaissuer_types.go [hunk 1]
```diff
@@ -36,7 +35,28 @@ type AWSPCAIssuerSpec struct {
...
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	// Namespace defines the space within which the secret name must be unique.
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
+	// Specifies the secret key where the AWS Access Key ID exists
+	// +optional
+	AccessKeyIDSelector SecretSelector `json:"accessKeyIDSelector,omitempty"`
+	// Specifies the secret key where the AWS Secret Access Key exists
+	// +optional
+	SecretAccessKeySelector SecretSelector `json:"secretAccessKeySelector,omitempty"`
```
Comment (RedbackThomson) [inline pkg/api/v1beta1/awspcaissuer_types.go:53]: Have you thought about including the `AWS_SESSION_TOKEN` key as well? It's a slim chance, but I could imagine that the values inside a secret are generated through an `sts assume-role`, and therefore would require this additional key.

File: pkg/api/v1beta1/awspcaissuer_types.go [hunk 2]
```diff
@@ -36,7 +35,28 @@ type AWSPCAIssuerSpec struct {
 	Region string `json:"region,omitempty"`
 	// Needs to be specified if you want to authorize with AWS using an access and secret key
 	// +optional
-	SecretRef v1.SecretReference `json:"secretRef,omitempty"`
+	SecretRef SecretReference `json:"secretRef,omitempty"`
+}
+
+type SecretReference struct {
```
Comment (RedbackThomson) [inline pkg/api/v1beta1/awspcaissuer_types.go:41]: My suggestion would be that this class composes the `SecretReference` type (https://pkg.go.dev/k8s.io/api/core/v1#SecretReference) and also is renamed.
```suggestion
type AWSCredentialsSecretReference struct {
	// Empty JSON tag is required to solve encountered struct field "" without JSON tag error.
	SecretReference `json:""`
```

File: pkg/api/v1beta1/awspcaissuer_types.go [hunk 3]
```diff
@@ -36,7 +35,28 @@ type AWSPCAIssuerSpec struct {
...
+	AccessKeyIDSelector SecretSelector `json:"accessKeyIDSelector,omitempty"`
+	// Specifies the secret key where the AWS Secret Access Key exists
+	// +optional
+	SecretAccessKeySelector SecretSelector `json:"secretAccessKeySelector,omitempty"`
+}
+
+// SecretSelector references a key within a Secret resource.
+type SecretSelector struct {
+	// The key of the entry in the Secret resource's `data` field to be used.
+	Key string `json:"key,omitempty"`
```
Comment (RedbackThomson) [inline pkg/api/v1beta1/awspcaissuer_types.go:59]: No need to define this yourself. There already exists a type in the K8s core library for `SecretKeySelector` https://pkg.go.dev/k8s.io/api/core/v1#SecretKeySelector

**Review by RedbackThomson — CHANGES_REQUESTED** [review]
Hey, great job on this. I can absolutely see the value in this addition, and your tests do a solid job of demonstrating that. I have left some inline comments about how we can re-use some existing k8s core types.

**Comment by anbaig** [comment]
Will take a look at PR after comments have been addressed

**Comment by kit837** [comment]
> Will take a look at PR after comments have been addressed

Hello @anbaig and @RedbackThomson -

Unfortunately @marshields is no longer working on this, and I'm not able to push to this branch.

I'm not sure of the best way to consolidate, but I opened up another PR to work on this feature: https://github.com/cert-manager/aws-privateca-issuer/pull/42

**Comment by RedbackThomson** [comment]
> > Will take a look at PR after comments have been addressed
> 
> Hello @anbaig and @RedbackThomson -
> 
> Unfortunately @marshields is no longer working on this, and I'm not able to push to this branch.
> 
> I'm not sure of the best way to consolidate, but I opened up another PR to work on this feature: #42

Okay, no problem. Hopefully we can close this one up, and I'll switch over to the other PR.

**Comment by laebshade** [comment]
Doesn't look like there's any way for me to personally close the PR; according to https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#marking-unfinished-pull-requests, it'll automatically close after 90 days.

**Pull request closed without merging**